### PR TITLE
Allow for whitespace differences between page text and text layer

### PR DESCRIPTION
**Pull request by robertknight**
When anchoring a quote in a PDF, the quote is first searched in text
extracted using PDF.js's `PDFPage.getTextContent` API, and the resulting
positions are used to create a range within the hidden text layer of a
page.

An issue we've seen several times when doing PDF.js upgrades is minor
changes to which spaces are included in the text layer. In the past
we've adapted our text extraction to match the text layer each time.
This slows down the process of upgrading PDF.js and makes maintaining
compatibility with a range of PDF.js releases more difficult. In the
most recent update, an `includeMarkedContent` option was added to the
`getTextContent` API, and the presence of that option could affect
whether certain whitespaces are included in the output or not [1].

Try to address this issue generally by mapping offsets from the page
text into offsets in the text layer in a way that ignores whitespace
differences.

A user-visible consequence of this change is that any leading or trailing whitespace in quotes
is not highlighted, due to the way character positions in the page text are translated
to character positions in the text layer.

 - Add `translateOffsets` utility, which translates a (start, end) pair
   of offsets in one string into offsets in another, ignoring certain
   characters in the strings.

 - Use `translateOffsets` utility in PDF anchoring to map quote offsets
   in the page text returned by `PDFPage.getTextContent` into offsets in
   the `textContent` of the text layer element.

Fixes https://github.com/hypothesis/client/issues/4331
Dependency of https://github.com/hypothesis/product-backlog/issues/1283

[1] https://github.com/hypothesis/browser-extension/pull/799#issuecomment-1079864595

### CODE REVIEW

File: src/annotator/anchoring/pdf.js [hunk 1]
```diff
@@ -187,21 +186,10 @@ function getPageTextContent(pageIndex) {
...
-
-    // Versions of PDF.js < v2.9.359 did not create elements in the text layer for
-    // text items that contained all-whitespace strings. Newer versions (after
-    // https://github.com/mozilla/pdf.js/pull/13257) do. The same commit also
-    // introduced the `hasEOL` property to text items, so we use the absence
-    // of this property to determine if we need to filter out whitespace-only strings.
-    const excludeEmpty = items.length > 0 && !('hasEOL' in items[0]);
-    if (excludeEmpty) {
-      items = items.filter(it => /\S/.test(it.str));
-    }
```
Comment (robertknight) [inline src/annotator/anchoring/pdf.js:202]: Since anchoring now tolerates differences between whitespace in the output of `pdfPage.getTextContent` vs the text layer, we don't need this compatibility switch.

File: src/annotator/anchoring/pdf.js [hunk 2]
```diff
@@ -295,21 +305,24 @@ async function anchorByPosition(pageIndex, start, end) {
...
-    const root = page.textLayer.textLayerDiv;
-
-    // Do a sanity check to verify that the page text extracted by `getPageTextContent`
-    // matches the transparent text layer.
     //
-    // See https://github.com/hypothesis/client/issues/3674.
-    if (pageText !== root.textContent) {
-      /* istanbul ignore next */
-      warnOnce(
-        'PDF text layer content does not match page text. This will cause anchoring misalignment.'
```
Comment (robertknight) [inline src/annotator/anchoring/pdf.js:307]: A variation on this warning will be re-introduced in a follow-up PR to handle cases where the page text and text layer content differ beyond just whitespace changes.

**Review by lyzadanger — COMMENTED** [review]
I've (embarrassingly!) spent a few hours reviewing this — parts of that was taking time to understand how PDF.js' text layer "works" — and I'm at the point where I think this would be more efficient by using voices to walk through where I'm getting bogged down: the internal implementation of `translateOffsets` and the supporting functions in `normalize`. Might be getting thrown by function/reference-naming and/or comment specifics and it might be something a simple conversation could clarify. I'm including the comments I came up with until this point...

Overall, my main input is that I want to be sure we define what "the text of the PDF page" means at any point in code and whose opinion matters most (the text content as retrieved through PDF.js' API or the textual content of the text layer `div`, e.g.?). It's feeling ambiguous. The use of `input` and `output` to name some of this is, oddly, further confusing me. Input to what? Output from where? :) I have a hunch there's some background context I'm simply missing.

I have my head around _most_ of what this is doing, but there's sort of some last-mile stuff, and some grokking of the sausage-making in `normalize` that I'd like to talk through together.

File: src/annotator/anchoring/pdf.js [hunk 3]
```diff
@@ -269,6 +256,29 @@ async function findPageByOffset(offset) {
   return { index: viewer.pagesCount - 1, offset: pageStartOffset, text };
 }
 
+/**
+ * Return true if `char` is an ASCII space.
+ *
+ * This is more efficient than `/\s/.test(char)` but does not handle Unicode
+ * spaces.
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/annotator/anchoring/pdf.js:263]: > Does this imply that encoding in text layers is ASCII-only (surely not?) or that we don't consider Unicode spaces to be whitespace (more likely?)? 

Your second try is right, we just treat non-ASCII spaces as not being spaces. That's acceptable here because we are trying to filter out the whitespace changes from one PDF.js version to another and, AFAIK, those differences are always in ASCII space chars.

> What's the efficiency difference here? 

According to some [ESBench tests](https://esbench.com/bench/624bf0486c89f600a57016f6), about 2-3x depending on the browser.

File: src/annotator/util/normalize.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+ *   // Input offsets specify the substring "bc" in the input. Result specifies
+ *   // the corresponding "b c" substring in the output.
+ *   translateOffsets('abcd', ' a b c d ', 1, 3, char => char !== ' ')
+ *
+ * @param {string} inputStr
+ * @param {string} outputStr
+ * @param {number} start - Start offset in `inputStr`
+ * @param {number} end - End offset in `inputStr`
+ * @param {(ch: string) => boolean} filter
+ * @return {[number, number]} - Start and end offsets in `outputStr`
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/annotator/util/normalize.js:63]: I honestly think it would be a bit weird if the function needed called with `translateOffsets(inputStr, outputStr, [start, end], filter)`. The result is a tuple simply because JS functions can only return a single value.

File: src/annotator/anchoring/pdf.js [hunk 5]
```diff
@@ -295,21 +305,24 @@ async function anchorByPosition(pageIndex, start, end) {
...
+    // will cause mis-anchoring.
 
-    const startPos = new TextPosition(root, start);
-    const endPos = new TextPosition(root, end);
+    const root = page.textLayer.textLayerDiv;
+    const textLayerStr = /** @type {string} */ (root.textContent);
+
+    const [textLayerStart, textLayerEnd] = translateOffsets(
+      pageText,
+      textLayerStr,
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/annotator/anchoring/pdf.js:318]: > I'm a little less clear on the provenance of the textContent of the text-layer root div. Presumably this is pulled from the PDF document's own text layer and chunked into positioned span elements. Are we to go on the assumption that the textContent of this div is reliably the textual content of the PDF? Does PDF.js do any manipulation or transformation of this textual content)? Put another way: is the textual content of this text-layer div subject to non-negligible change across PDF.js versions or do we see it as generally constant?

The `textContent` of the text layer root is ultimately constructed from the output of the same text extraction API that we use here, and the mapping of that text into spans is pretty direct. However what happened here, and could happen in future, is that PDF.js calls the text extraction API with different arguments than we do and that altered the whitespace in the output. Previously we tried to adapt our calls to the API so that we called the text extraction API with the same arguments and filtered the results in the same way that PDF.js does, but this complicates maintenance over the long term. 

To put it another way, I think of the contents of the text layer div as being a version of what `getPageTextContent` returns that has been corrupted by adding or removing whitespace in arbitrary locations.

> There are some places in the code that make unqualified references to "the page's text" (e.g. the docstring for this method, the start and end args).

References to "the page's text" generally refer to what `getPageTextContent` returns.

File: src/annotator/util/normalize.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+/**
+ * Translate a (start, end) pair of offsets in an input string to corresponding
+ * offsets in an output string, where positions in the input and output strings
+ * are related by counting the number of preceding characters that pass a
+ * filter.
+ *
+ * In typical usage, the output string is a corrupted version of the input
+ * string, where characters that do not match the filter (eg. whitespace)
+ * have been inserted or deleted at arbitrary locations.
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/annotator/util/normalize.js:47]: The terms input/output originate from before this was extracted into a generic utility. Perhaps it would be better to just call them "stringA" and "stringB" or something. The important idea is that stringB is a copy of stringA which has been modified by adding or removing whitespace (or some other characters determined by `filter`) at arbitrary locations. The reason that `filter` is an argument to this function is because different callers might have different ideas of what spaces (or other characters) may have been added or removed by the corruption process.

File: src/annotator/util/normalize.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+ * @param {number} end - End offset in `inputStr`
+ * @param {(ch: string) => boolean} filter
+ * @return {[number, number]} - Start and end offsets in `outputStr`
+ */
+export function translateOffsets(inputStr, outputStr, start, end, filter) {
+  const startCount = countChars(inputStr, filter, 0, start);
+  const startToEndCount = countChars(inputStr, filter, start, end);
+
+  // Find smallest offset with `startCount` matching chars before it in output,
+  // then adjust to the largest such offset.
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/annotator/util/normalize.js:70]: This is probably best explained with an example:

If the "input" string is "abc", the "output" string is "a b c" and the start/end offsets refer to the character "b" in the input string, then there are multiple results that would cover the same "b" character in the output (they could be the positions of just the "b" or include the space before/after it). This logic is for choosing between those possibilities by picking the largest start position and the smallest end position. In other words, trimming leading and trailing spaces from the output.

Does this make sense?

File: src/annotator/anchoring/pdf.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/anchoring/pdf.js:318]: > However what happened here, and could happen in future, is that PDF.js calls the text extraction API with different arguments than we do and that altered the whitespace in the output. Previously we tried to adapt our calls to the API so that we called the text extraction API with the same arguments and filtered the results in the same way that PDF.js does, but this complicates maintenance over the long term.

OK, this was really key. Is it possible to somehow get this context explained in comments or  commit metadata?

File: src/annotator/util/normalize.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/util/normalize.js:47]: I think `stringA` and `stringB` might be better. Is it more accurate to say "`stringB` is a copy of `stringA`" or that "`stringA` and `stringB` are strings representing the same textual content with potential differences in whitespace"? 

I also think the word "corrupted" in the function docstring is throwing me.

File: src/annotator/util/normalize.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/util/normalize.js:70]: > start/end offsets

Are we talking about one offset at a time or two here?

**Review by lyzadanger — APPROVED** [review]
Accidentally submitted last review as "comment" instead of "approve." Here's an approval!

File: src/annotator/util/normalize.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+ * @example
+ *   // Input offsets specify the substring "bc" in the input. Result specifies
+ *   // the corresponding "b c" substring in the output.
+ *   translateOffsets('abcd', ' a b c d ', 1, 3, char => char !== ' ')
+ *
+ * @param {string} inputStr
+ * @param {string} outputStr
+ * @param {number} start - Start offset in `inputStr`
+ * @param {number} end - End offset in `inputStr`
+ * @param {(ch: string) => boolean} filter
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/annotator/util/normalize.js]: Done. I added some comments here.

File: src/annotator/util/normalize.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+  const startToEndCount = countChars(input, filter, start, end);
+
+  // Find smallest offset with `startCount` matching chars before it in output,
+  // then adjust to the largest such offset.
+  let outputStart = advance(output, beforeStartCount, filter);
+  while (outputStart < output.length && !filter(output[outputStart])) {
+    ++outputStart;
+  }
+
+  const outputEnd = advance(output, startToEndCount, filter, outputStart);
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/annotator/util/normalize.js]: Thanks for the feedback. I added some comments along these lines in an additional commit.
